### PR TITLE
Updates to LIP-36 PR

### PR DESCRIPTION
**Pull request by yondonfu**
**What does this pull request do? Explain your changes. (required)**

[Overall diff](https://github.com/livepeer/protocol/compare/yf/cumulative-earnings-changes) against `streamflow` branch.

The BondingManager bytecode size is now 23280. The size can be calculated with the following steps:

```
npm run compile
cat build/contracts/BondingManager.json | jq '.deployedBytecode' | wc -m | xargs -I {} echo "{} / 2 - 1" | bc
```

**Specific updates (required)**

- Cherry-picked a75e2b8 from `streamflow` to avoid dealing with rebase conflicts for now
- Fixed failing test cases in #390 
- Re-use ` pendingStakeAndFees()` helper in `pendingStake()`, `pendingFees()` and `updateDelegatorWithEarnings()`
   - This also means that the pre-LIP36 earnings claiming algorithm will no longer update the reward/fee pool and claimable stake for each round and that rewards/fees in a pool that are unaccounted for due to a delegator claiming before those rewards/fees are added to the pool will no longer be re-distributed to other delegators that have not claimed from the pool yet. After looking at the code some more, I agree with @kyriediculous 's point that the advantages of this change outweigh the disadvantages
- Reduce the diff for EarningsPool
    - Use separate `updateCumulativeRewardFactor()` and `updateCumulativeFeeFactor()` functions from EarningsPoolLIP36
    - Remove deprecated functions `addToRewardPool()`, `addToFeePool()` and `claimShare()` from EarningsPool
- Remove unneeded storage update for activeCumulativeRewards in `updateTranscoderWithFees()`
- Updated ethlint config and fixed all warnings/errors
- Reverted a few temporary changes

May still be useful to add additional tests, but left that out for now.

**How did you test each of these updates (required)**

Ran test suite.

**Does this pull request close any open issues?**

N/A

**Checklist:**

- [x] README and other documentation updated
- [x] All tests using `npm run test` pass

### CODE REVIEW

File: contracts/bonding/BondingManager.sol
Comment (kyriediculous) [inline contracts/bonding/BondingManager.sol:814]: I understand why this was added to the end, it breaks clients still accessing tuples using numeric indexes. But most web3 libraries now also convert the tuples to objects with named so index based access is no longer needed. 

I felt like grouping lastRewardRound and lastFeeRound was a little clearer in terms of readability. 

Just to clarify why I added it in between, the tests that still used index based access were changed to used object keys instead. But I'll leave it up to you in terms what you think is the best tradeoff here (readability or old client compatibility)

File: test/integration/Earnings.js
Comment (kyriediculous) [inline test/integration/Earnings.js]: Can remove this

**Review by kyriediculous — COMMENTED** [review]
Did a first pass, going to pull in the branch to do a second pass in the vscode UI and run some testing on the actual state transitions. Looks good so far though !

**Comment by yondonfu** [comment]
f53c96a Fixes a failing test case related to updating activeCumulativeRewards in updateTranscoderWithFees
ab87bf6 Always sets lastFeeRound to the current round since its possible (although unlikely in practice due to the way the client is implemented) for a ticket to be redeemed with round N and then a ticket to be redeemed with round N - 1.

File: test/gas-report/poolsize.js
Comment (kyriediculous) [inline test/gas-report/poolsize.js]: This was re-commented (it wasn't prior to this commit) due to coverage being flaky on this ? https://github.com/livepeer/protocol/issues/393

File: test/unit/BondingManager.js
Comment (kyriediculous) [inline test/unit/BondingManager.js:7]: Looks like this was re-added by vscode plugins upon saving?  I don't mind leaving it there tbh as it's the actual dependency that's being injected anyway.

File: contracts/bonding/BondingManager.sol
Comment (kyriediculous) [inline contracts/bonding/BondingManager.sol]: I believe this is no longer correct after  the changes, it only serves the single purpose of initialising the correct `lastClaimRound` earningsPool for the next "epoch" (for lack of better terminology)

**Comment by kyriediculous** [comment]
Ran some tests and changes look good to me otherwise , just needs a `npm run eslint:fix` it seems.

File: test/integration/Earnings.js
Comment (yondonfu) [reply test/integration/Earnings.js]: Fixed in fcedeac

File: contracts/bonding/BondingManager.sol
Comment (yondonfu) [reply contracts/bonding/BondingManager.sol]: Fixed in 654de5b

File: test/gas-report/poolsize.js
Comment (yondonfu) [reply test/gas-report/poolsize.js]: Fixed in 654de5b

**Review by kyriediculous — APPROVED** [review]
LGTM pending green CI

**Pull request merged**
